### DISCUSSION TITLE
Help with VFD and max stepper current on BTT SKR 3 EZ (STM32H723 variant)

### DISCUSSION THREAD

**Discussion by ed-apex** [grblHAL/core#631] [Q&A]
Hi,

I'm currently doing a refit / upgrade to my cnc router.  I'm now using a BTT SKR 3 (STM32H723).  I compiled my own FW to implement the sdcard along with the 5160 drivers (the latter not seeming to work on the webbuilder based fws).  I have been successful in achieving all my goals and the machine is working just peachy except... NO SPINDLE!  I am sort of flummoxed as it seems to be configured correctly in both the fw and wired correctly.

I'm using a Huanyang HY series vfd.    It is tested and working from the control panel.  I have been using Huanyang1 from the spindles driver.  I have configured it for RS485 control with PD01 = 2, PD02 = 2, PD163 =1, PD164 = 2 and PD165 =3.  I have the settings in grbl configured suitably to match with the address set to 1, and the baudrate set to 19200.  I have not messed with the communication protocols as I believe they are 8N1 RTU by default.  I get an alarm 14 every time I try to start the spindle.  I have tried every combination of reversing the tx and rx pin to my ttl / serial to RS485 adapter along with the A and B outputs and the terminating resistor on and off.  Same result. I have used a web build without the trinamic drivers in the board config which allows iosender to communicate with the board and I still can't get the spindle to start.  Alarm 14.  PA 9 and PA 10 are tx and rx on the SKR 3 EZ and the $pins in every compilation I have done along with the successful non trinamic webbuilder implementations match that.  I'm at a loss here.  I actually need this machine for work (yes it is a hobby grade machine, but it has a specific use) and I'm getting pretty behind and am really stressed.

The other issue I'm having is I can't find a way in the source code to lift the current restrictions for the stepper motors.  The motors are rated for 2.8 amps and the 5160s in the packaging they are in with wrap around heatsinks and fans should be good for 3amps at 24v.  The ceiling is 1584 as allowed by grblhal.  I'd like to raise it.  If I'm not mistaken, RMS should be about 77% of the peak rating, so I'd like to be in the 2100ma range at least.  I may be off base on all of this, so input is appreciated. 

I'm drinking a bit from the fire hose hear trying to wrap this up.  The physical build is done, it's homing, autosquaring wonderfully. I just need my spindle working and a little more juice to the drivers.  I'd like to go rs485, but I'm about ready to go 0-10 with a pwm to analog adapter as I just need this running asap.  Thanks to all for the help in advance.  I'm so stressed and frustrated right now.  I know it's something I'm doing wrong and I just can't figure it out.  For reference, here's a link to the ttl / serial to rs485 adapter I'm using: https://www.amazon.com/dp/B0BRFRS6R9

**Comment by dresco** [comment]
Hi, thanks for reporting.

I've got the H743 variant of the SKR3 here, and also a Huanyang VFD. Will confirm settings tomorrow, but have definitely had this combination working on the bench..

Do you have access to any sort of logic analyser or scope?

Cheers, Jon.

**Comment by dresco** [comment]
@ed-apex I've set my SKR-3 back up with the Huanyang VFD. All seems to be working as expected here, what does your wiring look like? There are some old comments on SKR3 connections in the H7 repo [here](https://github.com/dresco/STM32H7xx/issues/6#issuecomment-1328084821).

**Comment by ed-apex** [comment]
I have a scope, no logic analyzer and a rs485 sniffer is on its way.  Here's a quick report on what my config is right now: 

$pins
[PIN:PA5,Feed hold]
[PIN:PA6,Cycle start]
[PIN:PA4,Emergency stop]
[PIN:PC13,Probe]
[PIN:PC1,X limit min]
[PIN:PC3,Y limit min]
[PIN:PC2,Y limit min 2]
[PIN:PC0,Z limit min]
[PIN:PA7,Aux in 0,P0]
[PIN:PD4,X step]
[PIN:PA15,Y step]
[PIN:PD15,Y2 step]
[PIN:PE2,Z step]
[PIN:PD3,X dir]
[PIN:PA8,Y dir]
[PIN:PD14,Y2 dir]
[PIN:PE3,Z dir]
[PIN:PD5,Motor CSX]
[PIN:PD0,Motor CSY]
[PIN:PE1,Motor CSZ]
[PIN:PC6,Motor CSM3]
[PIN:PD6,X enable]
[PIN:PD1,Y enable]
[PIN:PC7,Y enable]
[PIN:PE0,Z enable]
[PIN:PB4,Mist]
[PIN:PB3,Flood]
[PIN:PB0,Aux out 0,P0]
[PIN:PB5,Aux out 1,P1]
[PIN:PB6,Aux out 2,P2]
[PIN:PE15,MISO,Motor]
[PIN:PE13,MOSI,Motor]
[PIN:PE14,I2C SCK,Motor]
[PIN:PA10,RX,Modbus]
[PIN:PA9,TX,Modbus]
ok
$I
[VER:1.1f.20241120:]
[OPT:VNMZHSL2,100,1024,3,0]
[AXS:3:XYZ]
[NEWOPT:ENUMS,RT+,HOME,ES,SED,TMC=7,SD]
[FIRMWARE:grblHAL]
[SIGNALS:HSEP]
[NVS STORAGE:*FLASH]
[FREE MEMORY:206K]
[DRIVER:STM32H723]
[DRIVER VERSION:241127]
[BOARD:BTT SKR-3]
[AUX IO:1,3,0,0]
[PLUGIN:Trinamic v0.20]
[PLUGIN:MODBUS v0.16]
[PLUGIN:HUANYANG VFD v0.12]
[PLUGIN:SDCARD v1.14]
[PLUGIN:FS macro plugin v0.11]
ok

**Comment by ed-apex** [comment]
Probably not helpful, but here's a link to some images of the connection to the tft header, the rs485 converter (also getting power and gnd from the tft header) and the vfd itself :  https://share.icloud.com/photos/02fBMfgUvLxAgvlMAKGcorEoA

**Comment by ed-apex** [comment]
Here's my platformio.ini: 
```text
# BTT SKR 3 (H723 variant) with USB serial, SD card, and Trinamic 5160 driver support -- 128KB bootloader
[env:btt_skr_30_h723_tmc5160_bl128]
board = generic_stm32h723vg
board_build.ldscript = STM32H723VGTX_FLASH_BL.ld
build_flags = ${common.build_flags}
              ${usb_h723.build_flags}
              ${sdcard.build_flags}
  # See Inc/my_machine.h for options
  -Wl,-Map,output.map
  -D BOARD_BTT_SKR_30
  -D TRINAMIC_ENABLE=5160
  -D SPINDLE0_ENABLE=SPINDLE_HUANYANG1
  -D MODBUS_ENABLE=1
  -D SPINDLE_FEEDBACK_ENABLE=1
  -D RS274NGC_ENABLE=1
  -D REGEX_ENABLE=1
  -D Y_GANGED=1
  -D Y_AUTO_SQUARE=1
  -D SDCARD_ENABLE=1
  -D ESTOP_ENABLE=1
  -D SAFETY_DOOR_ENABLE=0

lib_deps = ${common.lib_deps}
           ${usb_h723.lib_deps}
           ${sdcard.lib_deps}
           motors
           #trinamic
           spindle
lib_extra_dirs = ${common.lib_extra_dirs}
                 ${usb_h723.lib_extra_dirs}
                 ${sdcard.lib_extra_dirs}
upload_protocol = dfu

**Comment by dresco** [comment]
The connections and config all look okay to me. (Always some ambiguity on TX/RX labelling, but as you say, you've tried both ways already).

VFD setup from your initial post matches my notes here;
```
PD001 = 2 RS485 spindle control
PD002 = 2 RS485 speed control
PD163 = 1 VFD address
PD164 = 2 19,200 baud
PD165 = 3 8N1 RTU
```
I believe the relevant grblHAL settings are just these (and I think are all defaults anyway);
```
ModBus baud rate	$374=3
ModBus RX timeout	$375=50
VFD ModBus address	$460=1
```

If you have a sender connected, I would expect to see data every second. Otherwise, a '?', or 'Ctrl-C' should send something, the TX and RX lights on my RS485 module flash with each data transfer.

As you have a scope I would suggest checking the signals next.

These Huanyang devices are only sort of Modbus RTU compatible, so a protocol decode would likely complain about invalid data. But really I'd just be looking for activity initially. I'd start with the logic level signals on the board, looking for TX and RX data. (You'll be getting the Alarm 14 because there is no valid response within the RX timeout period).

Be prepared to be horrified by the noise on the RS485 side - these VFDs are nasty! :) You should be able to see something if you can use a math function to subtract one channel from the other (or even better have differential probe, but not at all needed for basic checks).

**Comment by ed-apex** [comment]
So I got it working in between the responses by throwing -D SPINDLE_FEEDBACK_ENABLE=0 in my platformio.ini.  Probably a hack solution, but it's all working.  Do you have any thoughts on how I can lift the restriction on the current settings for the steppers?  I've tried a few things but all of them caused the firmware to not initialize.

**Reply by ed-apex** [reply]
Scratch that, it only works after flashing the fw after a full chip erase.  If I power cycle, start io sender, alarm 14 if I try to run the spindle.  I'm losing my mind on this!

**Comment by ed-apex** [comment]
Well... I did a fresh pull from your repo Jon and just did a clean sheet of paper build and now the spindle works reliably (YAYYY!).  So that's solved.  Also, the fancy wave share ttl to rs485 adapter works peachy and has a lot of nice features.  Any insights on upping my current level for my steppers.  I did see a lot in the source about it making current calculations based on feedback from the driver. I'm not sure if that means I'm getting the most out of my 5160 ez drivers or not.

**Reply by dresco** [reply]
That's good news, I was just about to reply and suggest double checking what you're seeing on the RX pin. I've previously seen spurious pulses out of the transceiver if there is too much noise on the RS485 side (dodgy cabling/grounding etc), which then messes up the received data.

Afraid I've no ideas on the Trinamic current limits, am sure @terjeio will though.

**Reply by ed-apex** [reply]
Looks like you can actually pass them a trinamic m906 command to set the current.  It's not persistent, but I can put it in the initialization script.  Thanks so much for putting this together for this driver.  It was the only board that would do the job I could get really short notice and I'm really happy with it.  I have to do a control retrofit on an aging Flexicam and I'm seriously considering using grblhal instead of my go to, Mach3.  I can't get over how compact the whole setup is.  Obviously the big machine has monster servo drives, but they take step and direction for commands and do all the encoder feedback position correcting internally. Maybe an SKR 3 is a viable replacement for a control cabinet full of 90s crap tinwisker growing electronics :)

**Reply by terjeio** [reply]
Current limit fixed in commit[ 47f5868](https://github.com/terjeio/Trinamic-library/commit/47f58689a01b9676590e736b8c0ef7e6b1c6ed9c).

**Comment by dresco** [comment]
Thanks, will update submodules in a bit...

**Comment by ed-apex** [comment]
Thank you both!  SKR 3 is a very good budget board option for grblhal with the 5160 drivers.  Especially with the ability to support 48v for the drivers.  Tough to beat for well under $100 all in.
